GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -53,70 +52,101 @@ public void Dispose()
             }
         }
 
-        private void TimerCallback(object unusedState)
+        private static EventMetadata CreateEventMetadata(Exception e)
         {
-            string errorMessage = null;
-            InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
-            IProductUpgrader productUpgrader;
-            bool deleteExistingDownloads = true;
-
-            if (ProductUpgraderFactory.TryCreateUpgrader(out productUpgrader, this.tracer, out errorMessage))
-            {
-                if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))
-                {
-                    deleteExistingDownloads = false;
-                }
-            }
-
-            if (errorMessage != null)
+            EventMetadata metadata = new EventMetadata();
+            if (e != null)
             {
-                this.tracer.RelatedError(errorMessage);
+                metadata.Add("Exception", e.ToString());
             }
 
-            if (deleteExistingDownloads)
-            {
-                ProductUpgraderInfo.DeleteAllInstallerDownloads();
-            }
+            return metadata;
         }
 
-        private bool TryDownloadUpgrade(IProductUpgrader productUpgrader, out string errorMessage)
+        private void TimerCallback(object unusedState)
         {
+            string errorMessage = null;
+
             using (ITracer activity = this.tracer.StartActivity("Checking for product upgrades.", EventLevel.Informational))
             {
-                Version newerVersion = null;
-                string detailedError = null;
-                if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                try
                 {
-                    return false;
-                }
+                    // The upgrade check always goes against GitHub
+                    GitHubUpgrader productUpgrader = GitHubUpgrader.Create(

```

Comment:
Do we expect the `GitHubUpgrader` instance to be the same each time the timer fires? Could we not keep this instance around for the lifetime of the `ProductUpgraderTimer` rather than each `TimerCallback(..)`?

Reply:
I suppose the configured ring could change after the service has started. If we went this route, changing the GVFS config would require a service restart to pick up the new values...

Reply:
> If we went this route, changing the GVFS config would require a service restart to pick up the new values...

Right, I suspected this might have been the reason for not doing it. Thanks for confirming! 😃